### PR TITLE
examples: custom load balancer example

### PR DESCRIPTION
**Pull request by temawi**
Example on how to implement a custom `LoadBalancer`.

### CODE REVIEW

File: examples/src/main/java/io/grpc/examples/customloadbalance/CustomLoadBalanceClient.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    LoadBalancerRegistry.getDefaultRegistry()
+        .register(new ShufflingPickFirstLoadBalancerProvider());
+
+    NameResolverRegistry.getDefaultRegistry().register(new ExampleNameResolverProvider());
+
+    String target = String.format("%s:///%s", exampleScheme, exampleServiceName);
+
+    logger.info("Use default first_pick load balance policy");
+    ManagedChannel channel = ManagedChannelBuilder.forTarget(target)
+        .usePlaintext()
```
Comment (temawi) [inline examples/src/main/java/io/grpc/examples/customloadbalance/CustomLoadBalanceClient.java:76]: use insecureChannelCredentials

File: examples/src/main/java/io/grpc/examples/customloadbalance/ShufflingPickFirstLoadBalancerProvider.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+public class ShufflingPickFirstLoadBalancerProvider extends LoadBalancerProvider {
+
+  private static final String RANDOM_SEED_KEY = "randomSeed";
+
+  @Override
+  public ConfigOrError parseLoadBalancingPolicyConfig(Map<String, ?> rawLoadBalancingPolicyConfig) {
+    Long randomSeed = null;
+    if (rawLoadBalancingPolicyConfig.containsKey(RANDOM_SEED_KEY)) {
+      randomSeed = ((Double) rawLoadBalancingPolicyConfig.get(RANDOM_SEED_KEY)).longValue();
```
Comment (temawi) [inline examples/src/main/java/io/grpc/examples/customloadbalance/ShufflingPickFirstLoadBalancerProvider.java:18]: try/catch for RuntimeException to be safer with reading the JSON

Comment (temawi) [reply examples/src/main/java/io/grpc/examples/customloadbalance/ShufflingPickFirstLoadBalancerProvider.java:18]: Mention taht the config can come from a remote source - be defensive

File: examples/src/main/java/io/grpc/examples/customloadbalance/ShufflingPickFirstLoadBalancer.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+  private final Helper helper;
+  private Subchannel subchannel;
+
+  /**
+   * This class defines the configuration used by this {@link LoadBalancer}. Note that no part of
+   * the gRPC library is aware of this interface, it will be populated by the
+   * {@link ShufflingPickFirstLoadBalancerProvider}.
+   */
+  static class ShufflingPickFirstLoadBalancerConfig {
```
Comment (temawi) [inline examples/src/main/java/io/grpc/examples/customloadbalance/ShufflingPickFirstLoadBalancer.java:52]: Rename to Config

File: examples/src/main/java/io/grpc/examples/customloadbalance/ShufflingPickFirstLoadBalancerProvider.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package io.grpc.examples.customloadbalance;
```
Comment (temawi) [inline examples/src/main/java/io/grpc/examples/customloadbalance/ShufflingPickFirstLoadBalancerProvider.java:1]: Copyright

File: examples/src/main/java/io/grpc/examples/customloadbalance/CustomLoadBalanceClient.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+      response = blockingStub.sayHello(request);
+    } catch (StatusRuntimeException e) {
+      logger.log(Level.WARNING, "RPC failed: {0}", e.getStatus());
+      return;
+    }
+    logger.info("Greeting: " + response.getMessage());
+  }
+
+
+  public static void main(String[] args) throws Exception {
```
[the remark this answers is not in the document]
Comment (temawi) [reply examples/src/main/java/io/grpc/examples/customloadbalance/CustomLoadBalanceClient.java:67]: The loadbalance server prints these out

File: examples/src/main/java/io/grpc/examples/customloadbalance/ShufflingPickFirstLoadBalancerProvider.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  private static final String RANDOM_SEED_KEY = "randomSeed";
+
+  @Override
+  public ConfigOrError parseLoadBalancingPolicyConfig(Map<String, ?> rawLoadBalancingPolicyConfig) {
+    Long randomSeed = null;
+
+    // The load balancing configuration generally comes from a remote source over the wire, be
+    // defensive when parsing it.
+    try {
+      if (rawLoadBalancingPolicyConfig.containsKey(RANDOM_SEED_KEY)) {
```
Comment (ejona86) [inline examples/src/main/java/io/grpc/examples/customloadbalance/ShufflingPickFirstLoadBalancerProvider.java:37]: This if no longer is necessary.

File: examples/src/main/java/io/grpc/examples/customloadbalance/ShufflingPickFirstLoadBalancerProvider.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        Object randomSeedObj = rawLoadBalancingPolicyConfig.get("randomSeed");
+        if (randomSeedObj instanceof Double) {
+          randomSeed = ((Double) randomSeedObj).longValue();
+        }
+      }
+    } catch (RuntimeException e) {
+      return ConfigOrError.fromError(
+          Status.UNAVAILABLE.withDescription("unable to parse LB config").withCause(e));
+    }
+    return ConfigOrError.fromConfig(new ShufflingPickFirstLoadBalancer.Config(randomSeed));
```
Comment (ejona86) [inline examples/src/main/java/io/grpc/examples/customloadbalance/ShufflingPickFirstLoadBalancerProvider.java:47]: Can we move this into the try-catch. The constructor could have a null-pointer check or similar.

**Review by ejona86 — APPROVED** [review]

File: examples/src/main/java/io/grpc/examples/customloadbalance/ShufflingPickFirstLoadBalancerProvider.java (hunk 7 above)
Comment (temawi) [reply examples/src/main/java/io/grpc/examples/customloadbalance/ShufflingPickFirstLoadBalancerProvider.java]: Done. Did not add a null check - it is ok for the config to have a null value for the random seed.

Comment (ejona86) [reply examples/src/main/java/io/grpc/examples/customloadbalance/ShufflingPickFirstLoadBalancerProvider.java]: I wasn't suggesting to make a null check. Just that "calling the constructor might also throw" (if not now, maybe in the future) and so we want to include it in the try-catch.

**Review by ejona86 — APPROVED** [review]

**Pull request merged**